### ISSUE TITLE
Cannot build TestStack.White solution with VS2015

### ISSUE THREAD

**Issue by jschiefer** [TestStack/White#326]
When building with VS2015 (Enterprise), the build of TestStack.White fails in the Todo.Core project, with the following exception:

```
Severity	Code	Description	Project	File	Line
Error		Fody: An unhandled exception occurred:
Exception:
Unknown custom metadata item kind: 6
StackTrace:
   at Microsoft.Cci.Pdb.PdbFunction.ReadCustomMetadata(BitAccess bits)
   at Microsoft.Cci.Pdb.PdbFunction..ctor(ManProcSym proc, BitAccess bits)
   at Microsoft.Cci.Pdb.PdbFunction.LoadManagedFunctions(BitAccess bits, UInt32 limit, Boolean readStrings)
   at Microsoft.Cci.Pdb.PdbFile.LoadFuncsFromDbiModule(BitAccess bits, DbiModuleInfo info, IntHashTable names, ArrayList funcList, Boolean readStrings, MsfDirectory dir, Dictionary`2 nameIndex, PdbReader reader)
   at Microsoft.Cci.Pdb.PdbFile.LoadFunctions(Stream read, Dictionary`2& tokenToSourceMapping, String& sourceServerData, Int32& age, Guid& guid)
   at Mono.Cecil.Pdb.PdbReader.PopulateFunctions()
   at Mono.Cecil.Pdb.PdbReader.ProcessDebugHeader(ImageDebugDirectory directory, Byte[] header)
   at Mono.Cecil.ModuleDefinition.ProcessDebugHeader()
   at Mono.Cecil.ModuleDefinition.ReadSymbols(ISymbolReader reader)
   at Mono.Cecil.ModuleReader.ReadSymbols(ModuleDefinition module, ReaderParameters parameters)
   at Mono.Cecil.ModuleReader.CreateModuleFrom(Image image, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(Stream stream, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
   at InnerWeaver.ReadModule() in c:\TeamCity\buildAgent\work\7495521761d392b9\FodyIsolated\ModuleReader.cs:line 23
   at InnerWeaver.Execute() in c:\TeamCity\buildAgent\work\7495521761d392b9\FodyIsolated\InnerWeaver.cs:line 28
Source:
Mono.Cecil.Pdb
TargetSite:
Void ReadCustomMetadata(Microsoft.Cci.Pdb.BitAccess)	Todo.Core		
```

Looking upstream, it appears that Fody had a range of issues related to the introduction of the Roslyn compiler infrastructure, that are now fixed:

https://github.com/Fody/Fody/issues/129
https://github.com/Fody/Fody/issues/141
https://github.com/Fody/Fody/issues/184
https://github.com/Fody/Fody/issues/187
https://github.com/Fody/Costura/issues/109
https://github.com/Fody/PropertyChanged/issues/126

I tried updating Fody to the latest version (1.29.3 as of this writing), and that solves the compilation issue. But I am too new to TestStack.White to convince myself that it didn't introduce any new problems...

**Comment by maxinfet** [comment]
I did the same thing when I attempted to build it in VS2015. All the issues with Fody were in the samples so I didn't really mess with it more after I got it to build. I haven't really had the time to mess around with the samples and see if they are working properly since I did that.

**Comment by JakeGinnivan** [comment]
This is something I can fix. Will push a fix within a few days.

Sent from my Windows Phone
________________________________

**Comment by jschiefer** [comment]
Awesome, thank you!

**Comment by Roemer** [comment]
Did this ever happen?

**Comment by maxinfet** [comment]
As far as I know this never happened, I was experiencing this issue when working on the nunit branch. I have the change in this branch https://github.com/TestStack/White/pull/300 but I need to break this down and the pull would be in the UIAComWrapper branch so it would be better if I just removed it entirely from there and brought it in through master. Also I don't really know what Fody is for so I wasn't able to check that the functionality was still working. I just checked that it compiled. I was hoping whoever reviewed the pull would be able to tell me if the newest version had any side affects.

**Comment by choagland** [comment]
I can comment that - having pulled down White and also having updating Fody in the sample projects - that the single UI test in the sample does pass. It successfully launches the "ToDo" application, enters a task, and then verifies that it worked. (On my machine I ported it over to an MS unit test instead of a NUnit test.)

To the degree that Fody does... something... it appears to be doing it properly, if you update it to 1.29.4, in the White solution. (You may need to fix the configuration xml after updating it.)

**Comment by JakeGinnivan** [comment]
I can add more context to this. Fody is a IL Weaver, it allows you to cross cut things after your assembly is compiled. 
White todo app is using it for INotifyPropertyChanged. PropertyChanged.Fody will take an automatic `get;set;` and weave property change events in. Like so:

```
public bool SomeProp { get; set; }
// to
public bool SomeProp
{
    get { return _someProp; }
    set
    {
        _someProp = value;
        OnPropertyChanged("SomeProp");
    }
}
```

As for why the old version doesn't work with 2015, VS2015 uses Roslyn as it's compiler. If you are not aware the old c++ compiler used to compile .net has been replaced with a new compiler written in c# which makes it much easier to add language features and such.
Roslyn fixed some long standing PDB metadata bugs in the old compiler, which in turn meant there was new flags within the PDB's and Mono.Cecil (another library which gives an API over the compiled IL of your assembly, kinda like reflection but waaay more powerful as it allows you to change stuff) could not read the new metadata.
Fody uses Mono.Cecil to do it's IL weaving, so when VS2015 builds, Fody tries to read the assembly using Mono.Cecil, which blows up so the build fails.

Upgrading to the newer versions of Fody/Mono.Cecil is really safe, I use different Fody weavers in the majority of applications I write.

Hope that helps :)

**Comment by maxinfet** [comment]
Great to hear that it's safe to make the upgrade. I will get a pull request to master put together when I get home.

**Comment by ivan-danilov** [comment]
I assume it was fixed in 1ff0b323a0790cff1cc08b2eb9cea1fd778e8ff0, which is in all important branches now. Closing issue.
